modules/seqexec/server/src/main/scala/seqexec/server/altair/AltairControllerEpics.scala

```diff
@@ -0,0 +1,321 @@
+
+  private def prepareMatrix(newPos: (Length, Length)): IO[Unit] = convertTcsAction(
+    epicsTcs.aoPrepareControlMatrix.setX(newPos._1.toMillimeters) *>
+      epicsTcs.aoPrepareControlMatrix.setY(newPos._2.toMillimeters)
+  )
+
+  implicit val fieldLensEq: Eq[FieldLens] = Eq.by(_.ordinal)
+
+  private def pauseNgsMode(fieldLens: FieldLens, currCfg: EpicsAltairConfig)(reasons: Set[Gaos.PauseCondition])
+  : IO[Set[ResumeCondition] => IO[Unit]] = {
+    val offsets = reasons.collectFirst { case OffsetMove(p, n) => (p, n) }
+    val newPos = offsets.map(Function.tupled(newPosition(currCfg)))
+    val newPosOk = newPos.forall(newPosInRange)
+    val matrixOk = newPos.forall(validateCurrentControlMatrix(currCfg, _)) && fieldLens =!= FieldLens.IN
+    val prepMatrixOk = newPos.forall(validatePreparedControlMatrix(currCfg, _)) && fieldLens =!= FieldLens.IN
+    val guideOk = !reasons.contains(GaosStarOff) //It can follow the guide star on this step
+
+    val needsToStop = !(newPosOk || matrixOk || guideOk)
+    val newCfg = (EpicsAltairConfig.guideStarCoords.modify(v => newPos.filter(_ => newPosOk).getOrElse(v)) >>>
+      EpicsAltairConfig.preparedMatrixCoords.modify(v => newPos.filter(_ => newPosOk && !matrixOk && !prepMatrixOk)
+        .getOrElse(v))) (currCfg)
+
+    if (needsToStop) {
+      convertTcsAction(epicsTcs.aoCorrect.setCorrections("OFF") *>
+        epicsTcs.targetFilter.setShortCircuit("Closed")) *>
+        newPos.filter(_ => newPosOk && !matrixOk && !prepMatrixOk).fold(IO.unit)(prepareMatrix) *>
+        convertTcsAction(epicsTcs.targetFilter.post) *>
+        IO(resumeNgsMode(newCfg))
+    }
+    else IO(dumbResume)
+  }
+
+  private val dumbResume = (_: Set[ResumeCondition]) => IO.unit
+
+  private def resumeNgsMode(currCfg: EpicsAltairConfig)(reasons: Set[Gaos.ResumeCondition]): IO[Unit] = {
+    val newPosOk = newPosInRange(currCfg.guideStarCoords)
+    val guideOk = reasons.contains(GaosStarOn)
+
+    if (newPosOk && guideOk) {
+      epicsAltair.waitMatrixCalc(CarStateGEM5.IDLE, 10.seconds) *>
+        convertTcsAction(epicsTcs.aoCorrect.setCorrections("ON") *>
+          epicsTcs.aoFlatten.mark *>
+          epicsTcs.targetFilter.setShortCircuit("Open") *>
+          epicsTcs.targetFilter.post.void
+        )
+    }
+    else IO.unit
+  }
+
+  private def checkStrapLoopState(currCfg: EpicsAltairConfig): TrySeq[Unit] =
+    currCfg.strapRTStatus.either(
+      SeqexecFailure.Unexpected("Cannot start Altair STRAP loop, RT Control status is bad."), ()
+    ) *>
+      currCfg.strapTempStatus.either(
+        SeqexecFailure.Unexpected("Cannot start Altair STRAP loop, Temperature Control status is bad."), ()
+      ) *>
+      currCfg.stapHVoltStatus.either(
+        SeqexecFailure.Unexpected("Cannot start Altair STRAP loop, HVolt status is bad."), ()
+      )
+
+  private def startStrapGate(currCfg: EpicsAltairConfig): IO[Unit] = ( IO.unit
+//    epicsAltair.strapGateControl.setGate(1) *>
```

Comment:
I guess this is for testing right?

Reply:
Those are the commands that are not working